File: scripts/run_multiparameter_case_study.py
#!/usr/bin/env python3
# SPDX-FileCopyrightText: (C) 2026 F. C. Langbein <frank@langbein.org>
# SPDX-FileCopyrightText: (C) 2026 S. P. O'Neil <sean.oneil@westpoint.edu>
# SPDX-FileCopyrightText: (C) 2026 S. Schirmer <s.m.shermer@gmail.com>
# SPDX-FileCopyrightText: (C) 2026 C. A. Weidner <c.weidner@bristol.ac.uk>
# SPDX-FileCopyrightText: (C) 2026 E. A. Jonckheere <jonckhee@usc.edu>
#
# SPDX-License-Identifier: AGPL-3.0-or-later
"""Multi-parameter and time-varying margins for the paper-2 case study.

Per controller (p = 3 joint structures H0, H1, H2):
  - certified safe-polytope radii (per axis, l2 and linf inradii);
  - directional margins along the 6 axis and 8 diagonal directions
    (Euclidean-normalised), with margin_tol brackets;
  - uniform time-varying margins r0 per structure and joint;
  - optionally (--adversary N) time-varying brackets [r_FS, m_adv] for
    the first N controllers on structure H1, together with the
    constant-class bracket [M_const, M_const_upper] at the same
    controller, structure and threshold, and the constancy-gap endpoints
    formed from them. Everything the gap needs sits in one record, so no
    later join can pair margins computed to different tolerances.

Writes results/multiparameter-margin-python/multiparam_<FT>.csv and
tv_bracket_<FT>.csv.
"""

from __future__ import annotations

import argparse
import csv
from pathlib import Path

import numpy as np

from qrobustness import (
    dH_structure,
    iterative_margin,
    make_fidelity_fn,
)
from qrobustness import multiparam as mp
from qrobustness import timevarying as tv

from _drivers import DEFAULT_FT, DEFAULT_MAX_ERROR, load_ensemble, three_structure_specs

ROOT = Path(__file__).resolve().parents[1]
OUT_DIR = ROOT / "results/multiparameter-margin-python"
FT = DEFAULT_FT
MARGIN_TOL = 1e-8
STRUCTURES = ("H0", "H1", "H2")


def main() -> None:
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--out", type=Path, default=OUT_DIR)
    ap.add_argument("--FT", type=float, default=FT)
    ap.add_argument("--max-error", type=float, default=DEFAULT_MAX_ERROR)
    ap.add_argument(
        "--controllers",
        type=int,
        default=0,
        help="limit to the first n controllers (0 = all)",
    )
    ap.add_argument(
        "--adversary",
        type=int,
        default=0,
        help="run the time-varying adversary for the first n controllers",
    )
    ap.add_argument(
        "--step",
        choices=("lipschitz", "angular"),
        default="lipschitz",
        help="Directional continuation rule: the Lipschitz "
        "surplus step (default; reproduces the historical "
        "output byte-for-byte) or the dominating Choi-angular "
        "step (same crossings, fewer evaluations; writes "
        "multiparam_<FT>_angular.csv)",
    )
    args = ap.parse_args()
    ft = args.FT

    problem, controllers = load_ensemble(max_error=args.max_error)
    if args.controllers:
        controllers = controllers[: args.controllers]

    directions = np.vstack([mp.axis_directions(3), mp.diagonal_directions(3)])
    dir_names = (
        [f"+e{j}" for j in range(3)]
        + [f"-e{j}" for j in range(3)]
        + [
            "diag" + "".join("p" if s > 0 else "m" for s in d * np.sqrt(3))
            for d in mp.diagonal_directions(3)
        ]
    )

    rows = []
    tv_rows = []
    for ci, c in enumerate(controllers):
        dt = c["tf"] / c["tau"]
        specs = three_structure_specs(problem, c)
        C, L = mp.structure_constants(specs, dt, c["tau"], ft, problem["dim"])
        fn = mp.make_multiparam_fidelity_fn(
            problem["H0"],
            problem["H1"],
            problem["H2"],
            c["u1"],
            c["u2"],
            problem["Uf"],
            dt,
            STRUCTURES,
        )
        F0 = fn(np.zeros(3))
        P = mp.safe_polytope(np.zeros(3), L, F0, ft)

        row = {
            "controller": ci + 1,
            "fid": c["fid"],
            "err": c["error"],
            "L_H0": L[0],
            "L_H1": L[1],
            "L_H2": L[2],
            "poly_r_H0": P.axis_radii[0],
            "poly_r_H1": P.axis_radii[1],
            "poly_r_H2": P.axis_radii[2],
            "inradius_linf": P.inradius_linf,
            "inradius_l2": P.inradius_l2,
            "r0_joint": tv.uniform_margin(L, F0, ft),
        }
        for j, tag in enumerate(STRUCTURES):
            row[f"r0_{tag}"] = tv.uniform_margin(L[j], F0, ft)
        dHs = [
            [problem["H0"]] * c["tau"],
            [c["u1"][k] * problem["H1"] for k in range(c["tau"])],
            [c["u2"][k] * problem["H2"] for k in range(c["tau"])],
        ]
        G = mp.joint_gauge(dHs, dt)
        AG = mp.angular_gauge(dHs, dt) if args.step == "angular" else None
        for d, name in zip(directions, dir_names, strict=True):
            # Sharper directional constant of the joint gauge
            # (Theorem gauge); the separable sum_j L_j |d_j| is the
            # fallback relaxation.  With --step angular the iteration
            # steps and certifies with the dominating Choi-angular
            # radius instead.
            L_dir = G.L_dir(d, ft, problem["dim"])
            res = mp.directional_margin(
                fn,
                L,
                ft,
                d,
                margin_tol=MARGIN_TOL,
                L_dir=L_dir,
                angular_gauge=AG,
                return_diagnostics=True,
            )
            row[f"M_{name}"] = res.M
            row[f"Mupper_{name}"] = res.M_upper
            row[f"nev_{name}"] = res.n_evals
        rows.append(row)
        print(
            f"controller {ci + 1}/{len(controllers)}  "
            f"inradius_l2={row['inradius_l2']:.3e}  "
            f"M_worst={min(row[f'M_{n}'] for n in dir_names):.3e}",
            flush=True,
        )

        if ci < args.adversary:
            tau = c["tau"]
            H_list = [
                problem["H0"] + c["u1"][k] * problem["H1"] + c["u2"][k] * problem["H2"]
                for k in range(tau)
            ]
            Hhat = dH_structure(
                problem["H0"], problem["H1"], problem["H2"], c["u1"], c["u2"], "H1"
            )
            scalar_fn = make_fidelity_fn(
                problem["H0"],
                problem["H1"],
                problem["H2"],
                c["u1"],
                c["u2"],
                problem["Uf"],
                dt,
                "H1",
            )
            F0_scalar = scalar_fn(0.0)
            r0 = tv.uniform_margin(L[1], F0_scalar, ft)
            # The constancy gap needs BOTH ends of the constant-class
            # bracket, so the scalar margin is run with the same
            # tolerance as the directional ones. M_upper is finite only
            # when an unsafe constant perturbation was actually
            # evaluated; a domain edge or an exhausted search leaves it
            # infinite, and the gap's upper endpoint is then unavailable
            # rather than clipped to something finite.
            res_const = iterative_margin(scalar_fn, L[1], ft, margin_tol=MARGIN_TOL)
            M = res_const.M
            M_upper = res_const.M_upper
            # The trajectory lower certificate for the same controller,
            # structure and threshold, computed here rather than joined
            # in later from a separate run.
            fsm = tv.fs_margin(Hhat, dt, F0_scalar, ft, r0)
            br = tv.adversarial_upper_bound(
                H_list,
                Hhat,
                dt,
                problem["Uf"],
                ft,
                r0,
                2.0 * M,
                rel_tol=2e-2,
                seed=1000 + ci,
            )
            # An upper witness exists only if the adversary exhibited a
            # violating trajectory; adversarial_upper_bound returns
            # delta_adv exactly then.
            adv_violated = br.delta_adv is not None
            gap_lo = max(0.0, M - br.m_adv) if adv_violated else None
            gap_hi = (M_upper - fsm.r_fs) if np.isfinite(M_upper) else None
            if gap_lo is not None and gap_hi is not None and gap_lo > gap_hi:
                raise SystemExit(
                    f"ERROR: controller {ci + 1} structure H1 at FT={ft:g} gives "
                    f"gap_lower={gap_lo:.6e} > gap_upper={gap_hi:.6e}. The two "
                    "endpoints come from inconsistent records (M_const/m_adv "
                    "against M_const_upper/r_FS); the inputs must be fixed "
                    "rather than the interval clipped."
                )
            tv_rows.append(
                {
                    "controller": ci + 1,
                    "structure": "H1",
                    "FT": ft,
                    "r0": br.r0,
                    "r_fs": fsm.r_fs,
                    "M_const": M,
                    "M_const_upper": M_upper,
                    "m_adv": br.m_adv,
                    "F_at_adv": br.F_at_adv,
                    "adv_violated": int(adv_violated),
                    "gap_lower": "" if gap_lo is None else gap_lo,
                    "gap_upper": "" if gap_hi is None else gap_hi,
                    "n_adversary_calls": br.n_evals,
                }
            )
            print(
                f"  tv bracket H1: r_FS={fsm.r_fs:.3e} <= M_tv <= {br.m_adv:.3e}"
                f"  (M_const in [{M:.3e}, {M_upper:.3e}]; constancy gap in ["
                + ("n/a" if gap_lo is None else f"{gap_lo:.3e}")
                + ", "
                + ("n/a" if gap_hi is None else f"{gap_hi:.3e}")
                + "])",
                flush=True,
            )

    out = args.out
    out.mkdir(parents=True, exist_ok=True)
    suffix = "_angular" if args.step == "angular" else ""
    path = out / f"multiparam_{ft:g}{suffix}.csv"
    with path.open("w", newline="") as f:
        w = csv.DictWriter(f, fieldnames=list(rows[0].keys()), lineterminator="\n")
        w.writeheader()
        w.writerows(rows)
    print(f"Wrote {path}")
    if tv_rows:
        path = out / f"tv_bracket_{ft:g}.csv"
        with path.open("w", newline="") as f:
            w = csv.DictWriter(
                f, fieldnames=list(tv_rows[0].keys()), lineterminator="\n"
            )
            w.writeheader()
            w.writerows(tv_rows)
        print(f"Wrote {path}")


if __name__ == "__main__":
    main()

File: matlab/examples/run_multiparameter_case_study.m
% SPDX-FileCopyrightText: (C) 2026 F. C. Langbein <frank@langbein.org>
% SPDX-FileCopyrightText: (C) 2026 S. P. O'Neil <sean.oneil@westpoint.edu>
% SPDX-FileCopyrightText: (C) 2026 S. Schirmer <s.m.shermer@gmail.com>
% SPDX-FileCopyrightText: (C) 2026 C. A. Weidner <c.weidner@bristol.ac.uk>
% SPDX-FileCopyrightText: (C) 2026 E. A. Jonckheere <jonckhee@usc.edu>
%
% SPDX-License-Identifier: AGPL-3.0-or-later
function run_multiparameter_case_study(varargin)
%RUN_MULTIPARAMETER_CASE_STUDY xQRM joint-parameter margin case study.
%   Peer of scripts/run_multiparameter_case_study.py.  It writes the same
%   CSV schema, with a language-qualified result directory selected by the
%   caller.  The angular stepping rule is the default paper protocol.

    p = inputParser;
    addParameter(p, 'root', '');
    addParameter(p, 'out', '');
    addParameter(p, 'FT', 0.999);
    addParameter(p, 'max_error', 1e-4);
    addParameter(p, 'controllers', 0);
    addParameter(p, 'step', 'angular');
    parse(p, varargin{:});
    opt = p.Results;

    % Bracket refinement, named to match the Python reference's MARGIN_TOL.
    margin_tol = 1e-8;

    if isempty(opt.root)
        here = fileparts(mfilename('fullpath'));
        opt.root = fileparts(fileparts(here));
    end
    addpath(fullfile(opt.root, 'matlab'));
    if isempty(opt.out)
        opt.out = fullfile(opt.root, 'results', 'multiparameter-margin-matlab');
    end
    if ~exist(opt.out, 'dir'); mkdir(opt.out); end

    ctrl = fullfile(opt.root, 'data', 'controllers', ...
        'problem9_tf15_K32_quasi-newton');
    problem = qrobustness.load_problem(fullfile(ctrl, 'problem9.mat'));
    controllers = qrobustness.load_controllers(fullfile(ctrl, 'controllers.csv'), ...
        opt.max_error);
    if opt.controllers > 0
        controllers = controllers(1:min(opt.controllers, numel(controllers)));
    end

    % Three uncertainty parameters: drift and the two controls. Named
    % once here as in the Python peer, which calls it N_PARAMS.
    n_params = 3;
    directions = [qrobustness.multiparam.axis_directions(n_params); ...
                  qrobustness.multiparam.diagonal_directions(n_params)];
    names = {'+e0', '+e1', '+e2', '-e0', '-e1', '-e2', ...
             'diagmmm', 'diagpmm', 'diagmpm', 'diagppm', ...
             'diagmmp', 'diagpmp', 'diagmpp', 'diagppp'};
    headers = {'controller', 'fid', 'err', 'L_H0', 'L_H1', 'L_H2', ...
        'poly_r_H0', 'poly_r_H1', 'poly_r_H2', 'inradius_linf', ...
        'inradius_l2', 'r0_joint', 'r0_H0', 'r0_H1', 'r0_H2'};
    for j = 1:numel(names)
        headers{end + 1} = ['M_' names{j}]; %#ok<AGROW>
        headers{end + 1} = ['Mupper_' names{j}]; %#ok<AGROW>
        headers{end + 1} = ['nev_' names{j}]; %#ok<AGROW>
    end
    values = zeros(numel(controllers), numel(headers));

    for ci = 1:numel(controllers)
        c = controllers{ci};
        dt = c.tf / c.tau;
        specs = {{'drift', problem.H0}, {'control', problem.H1, c.u1}, ...
                 {'control', problem.H2, c.u2}};
        [~, L] = qrobustness.multiparam.structure_constants( ...
            specs, dt, c.tau, opt.FT, problem.dim);
        Hhat_lists = {repmat({problem.H0}, 1, c.tau), ...
            arrayfun(@(u) u * problem.H1, c.u1, 'UniformOutput', false), ...
            arrayfun(@(u) u * problem.H2, c.u2, 'UniformOutput', false)};
        joint = qrobustness.multiparam.joint_gauge(Hhat_lists, dt);
        angular = qrobustness.multiparam.angular_gauge(Hhat_lists, dt);
        fidelity_fn = @(mu) multiparam_fidelity(problem, c, dt, mu);
        F0 = fidelity_fn(zeros(3, 1));
        poly = qrobustness.multiparam.safe_polytope(zeros(3, 1), L, F0, opt.FT);

        row = [ci, c.fid, c.error, L(:).', poly.axis_radii(:).', ...
            poly.inradius_linf, poly.inradius_l2, ...
            qrobustness.timevarying.uniform_margin(L, F0, opt.FT), ...
            qrobustness.timevarying.uniform_margin(L(1), F0, opt.FT), ...
            qrobustness.timevarying.uniform_margin(L(2), F0, opt.FT), ...
            qrobustness.timevarying.uniform_margin(L(3), F0, opt.FT)];
        for di = 1:size(directions, 1)
            d = directions(di, :).';
            kwargs = {'margin_tol', margin_tol, 'return_diagnostics', true, ...
                'L_dir', qrobustness.multiparam.joint_gauge_L_dir( ...
                    joint, d, opt.FT, problem.dim)};
            if strcmpi(opt.step, 'angular')
                kwargs = [kwargs, {'angular_gauge', angular}]; %#ok<AGROW>
            end
            result = qrobustness.multiparam.directional_margin( ...
                fidelity_fn, L, opt.FT, d, kwargs{:});
            row = [row, result.M, result.M_upper, result.n_evals]; %#ok<AGROW>
        end
        values(ci, :) = row;
        fprintf('controller %d/%d inradius_l2=%.3e\n', ...
            ci, numel(controllers), poly.inradius_l2);
    end

    suffix = '';
    if strcmpi(opt.step, 'angular'); suffix = '_angular'; end
    path = fullfile(opt.out, sprintf('multiparam_%g%s.csv', opt.FT, suffix));
    write_csv(path, headers, values);
    fprintf('Wrote %s\n', path);
end

function F = multiparam_fidelity(problem, controller, dt, mu)
    % Fidelity at a joint perturbation mu of the three structures,
    % multiplicative on each as in the main study.
    H = cell(1, controller.tau);
    for k = 1:controller.tau
        H{k} = (1 + mu(1)) * problem.H0 ...
            + (1 + mu(2)) * controller.u1(k) * problem.H1 ...
            + (1 + mu(3)) * controller.u2(k) * problem.H2;
    end
    F = qrobustness.gate_fidelity(qrobustness.propagator(H, dt), problem.Uf);
end

function write_csv(path, headers, values)
    % Write a numeric table with an LF line terminator and 16 significant
    % digits, matching what the Python peer's DictWriter produces.
    fid = fopen(path, 'w');
    cleaner = onCleanup(@() fclose(fid));
    fprintf(fid, '%s\n', strjoin(headers, ','));
    fmt = [repmat('%.16g,', 1, numel(headers) - 1), '%.16g\n'];
    for row = 1:size(values, 1)
        fprintf(fid, fmt, values(row, :));
    end
end
